webclient/src/agave.ts

```diff
@@ -34,69 +35,77 @@ export class AgaveClient {
     this.onOpen = onOpen;
     this.onJson = onJson;
     this.onImage = onImage;
+    this.cb = new CommandBuffer();
+    this.sessionName = "";
+    this.url = url + "?mode=" + rendermode;
+    this.socket = undefined;
+  }
 
-    // First order of business: connect!
-    this.socket = new WebSocket(url + "?mode=" + rendermode);
-
-    // set binarytype according to how we expect clients to deal with received image data
-    this.socket.binaryType = "blob"; //"arraybuffer";
-
-    // do some stuff on initial connection
-    this.socket.onopen = (_ev: Event) => {
-      this.setResolution(256, 256);
-      // put agave in streaming mode from the get-go
-      this.streamMode(1);
-      this.flushCommandBuffer();
-
-      // user provided callback
-      if (this.onOpen) {
-        this.onOpen();
-      }
-    };
-
-    // TODO - handle this better
-    this.socket.onclose = (_ev: CloseEvent) => {
-      setTimeout(function () {
-        console.warn("connection failed. refresh to retry.");
-      }, 3000);
-    };
-
-    // handle incoming messages
-    this.socket.onmessage = (evt: MessageEvent<unknown>) => {
-      if (typeof evt.data === "string") {
-        const returnedObj = JSON.parse(evt.data);
-        if (returnedObj.commandId === COMMANDS.LOAD_DATA[0]) {
-          console.log(returnedObj);
-          // let users do something with this data
-          if (this.onJson) {
-            this.onJson(returnedObj);
-          }
+  async connect() {
+    return new Promise((resolve, reject) => {
+      // First order of business: connect!
+      this.socket = new WebSocket(this.url);
+
+      // set binarytype according to how we expect clients to deal with received image data
+      this.socket.binaryType = "blob"; //"arraybuffer";
+
+      // do some stuff on initial connection
+      this.socket.onopen = (_ev: Event) => {
+        this.setResolution(256, 256);
+        // put agave in streaming mode from the get-go
+        this.streamMode(1);
+        this.flushCommandBuffer();
+
+        // user provided callback
+        if (this.onOpen) {
+          this.onOpen();
         }
-        return;
-      }
 
-      const arraybuf = evt.data;
-      // let users do something with this data
-      if (this.onImage) {
-        this.onImage(arraybuf as Blob);
-      }
-    };
+        resolve(this);
+      };
+
+      // TODO - handle this better
+      this.socket.onclose = (_ev: CloseEvent) => {
+        setTimeout(function () {
+          console.warn("connection failed. refresh to retry.");
+        }, 3000);
+      };
+
```

Comment:
Would you want this code to have a try/catch that tries to reestablish connection before asking the user to refresh?

Reply:
for now there aren't really any known use cases where this should happen.  This is really just some placeholder code (there's a TODO above). I'll simplify it even more.